Show interesting currently running matches

Rhacco opened the pull request:
See #7 

Added a new Activity that shows overviews of the top live games fetched from the official Dota 2 API (compare [this](https://wiki.teamfortress.com/wiki/WebAPI/GetTopLiveGame)). The returned result seems to include current highest average MMR and relevant tournament matches (decided by Valve) and always returns 10 in total.

Rhacco committed:
New Activity to show live matches later and (according) refactoring

Rhacco committed:
Tons of code only to fetch and display Steam server IDs of 'top live games'

Rhacco committed:
Trying to fetch and use Dota 2 'realtime stats', not quite working yet

EdbertChan committed:
Show live matches fix to list

Rhacco committed:
Merge pull request #8 from EdbertChan/fix/showlivematches

Show live matches fix to list

Rhacco committed:
Now fetching and showing match IDs of 'top live games' properly, thank you @EdbertChan

Rhacco committed:
Ditched database for realtime data as it updates frequently

Rhacco committed:
Refactored response models and added new fields

Rhacco committed:
Fixed broken Design mode in AS, renamed view model

Rhacco committed:
Now displaying 'top live games' with players

Rhacco committed:
Added new screenshots

Rhacco committed:
Updated README.md

Rhacco committed:
Some final refactoring

- Removed some classes not needed/used for querying live match data
- Improved the checks for an active internet connection

Rhacco committed:
Fixed introduced possible NullPointerException

Turns out a simple careless cast can lead to a NullPointerException.
Still have to be very careful even with Kotlin's null safety measures it
seems

Rhacco committed:
Some more final refactoring

- Updated an inaccurate comment
- Improved checks for network connection again. This code structure
seems pretty good for any API service we're going to implement
- Introduced Xdroid with its Toaster module for convenient toasting

Rhacco committed:
Now using Xdroid's Toaster module for all toasts

Rhacco committed:
Updated a log message and the list adapter

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/activities/LiveMatchesActivity.kt:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    private fun updateLiveMatches() {
+        val disposable = getDota2OfficialAPIService()
```
Why are we directly exposing the reactive strucutre? Why aren't we using a viewmodel for this?

Showing the reactive structure completely defeats the purpose of the viewmodel architecture

Rhacco replied:
Fixed

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/activities/LiveMatchesActivity.kt:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    private fun updateListEntry(serverSteamId: Long, averageMMR: Int) {
+        val disposable = getDota2OfficialAPIService()
```
Same as above here.

Any advanced manipulation of the format of the response should be done on the repository level. By advanced manipulation, I mean text formatting of the response or the conversion of the response to another data structure (ie array to LinkedList) or something.

The activity should be more or less responsible for seeing a list of objects and throwing that to the adapter.

Databinding with the XML can also be used to display a certain string(s) in a particular order. Consider:


<layout>
<data>
        <variable
            name="match"
            type="foo.bar.Match" />
</data>

<TextView
android:id="@+id/player0
android:src="@{match.player[0]}"
/>
....
</layout>

String manipulation in the activity is entirely inappropriate.

Rhacco replied:
So from what you've told me I thought the ViewModel > Repository architecture only applies to queries with a database in the background since you said DAO + API = Repository. I don't use a DAO/database here since live data changes frequently, only the direct query on the API, so I considered the extra classes superfluous and got rid of them.

Right now I'm seeing "The ViewModel class is designed to store and manage UI-related data so that the data survives configuration changes such as screen rotations" in official Android docs so I definitely see the point of introducing this architecture again. I think storing the live data in memory in a new LocalDataSource is the way to go (still no database but with the ViewModel architecture), so I'm going with this now and let you take another look at it

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/lists/LiveMatchesListAdapter.kt:
```diff
@@ -0,0 +1,48 @@
+import android.widget.TextView
+import github.com.rhacco.dota2androidapp.R
+
+class LiveMatchesListAdapter(context: Context) : BaseAdapter() {
```
We should be using RecyclerView.Adapter, not BaseAdapter. Recyclerview already implements this recycling/viewholding pattern for reuse.

Rhacco replied:
Done

EdbertChan commented on app/src/main/res/layout/activity_live_matches.xml:
```diff
@@ -0,0 +1,11 @@
+              android:layout_width="match_parent"
+              android:layout_height="match_parent"
+              android:orientation="vertical" android:paddingLeft="10dp" android:paddingRight="10dp">
+    <ListView
```
DONT USE LISTVIEW. Use RecyclerView. 99% of the time, RecyclerView is more optimal, especially in this case where you're reusing the same layout anyway for each livematch.

Rhacco replied:
So in the the discussion [here](https://stackoverflow.com/questions/28392554/should-we-use-recyclerview-to-replace-listview) they say "If ListView works for you, there is no reason to migrate." I guess it sucks a bit that I wasn't referred to RecyclerView directly when searching for how to create a simple list, but I don't really see a good reason to put more effort into this again to migrate to RecyclerView. Especially since I'm planning to address UI devs in the Discord server later to give them a chance to gain more experience rewriting parts of this app's UI.

Rhacco replied:
Decided to migrate to RecyclerView despite my original comment on this

EdbertChan commented on app/src/main/res/layout/list_live_matches.xml:
```diff
@@ -0,0 +1,31 @@
+                  android:layout_height="match_parent"
+                  android:orientation="horizontal"
+                  android:weightSum="2">
+        <TextView
```
Use databinding. Take a look at it and add a textview for each player. It will help you avoid the aforementioned string manipulation in the activity.

Rhacco replied:
Databinding sounds like something we might want to use eventually. I don't want to introduce it in this PR as another new feature. Also it seems to have some drawbacks seen [here](https://medium.com/@Miqubel/4-reasons-im-not-using-android-data-binding-e62127c2650c) for example

EdbertChan replied:
Yes, databinding does have these features but we are unlikely to run into them on this screen particularly. This app is not complex enough to worry about scalability or the spahgetti code issue.

I do like the idea of using butterknife (although I don't see you using it) but I have never used it in tandem with a listview, only with recycler.

I would suggest reworking this to have a layout such as:

<Textview
id="@+id/player1"/>
<TextView
id="@+id/player2/>

Rhacco replied:
I'm using the Kotlin Android Extensions (basically Butter Knife for Kotlin) for convenient data binding now. Also reworked the layout to have an individual TextView for each player

EdbertChan commented:
If you're going to refactor the location of these files, please do that in a separate request. Refactoring of all project files and adding a new feature is very confusing and adds a lot of work for the reviewers.

Rhacco committed:
Applied the ViewModel architecture to TopLiveGames

Rhacco committed:
Applied ViewModel architecture to RealtimeStats

Rhacco committed:
Updated screenshot and README.md

Rhacco committed:
Moved string manipulation out of Activity

... into the related ViewModel. This seemed like the most reasonable
approach for now since the Repository should probably return the
response mostly unformatted so it can be reused elsewhere. Also, live
matches now show tournament matches first and then ranked matches sorted
by average MMR (descending).

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/viewmodel/MatchesViewModel.kt:
```diff
@@ -6,49 +6,85 @@ import android.arch.lifecycle.MediatorLiveData
+                                            listItemData.mPlayersRadiant =
+                                                    listItemData.mPlayersRadiant + "\n" + it
+                                        }
+                                result.teams[1].players
```
DRY (Don't repeat yourself). You can do this with result.teams[i] looping.

Rhacco replied:
Fixed

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/viewmodel/MatchesViewModel.kt:
```diff
@@ -6,49 +6,85 @@ import android.arch.lifecycle.MediatorLiveData
+                                        }
+                                        .forEach {
+                                            listItemData.mPlayersDire =
+                                                    listItemData.mPlayersDire + "\n" + it
```
Ew. Please see my comment on the layout about removing this. You absolutely do not need to do this \n

Regardless if you choose to use databinding or not, please do this sort of binding/formatting in the adapter.

Rhacco replied:
Fixed

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/viewmodel/MatchesViewModel.kt:
```diff
@@ -6,49 +6,85 @@ import android.arch.lifecycle.MediatorLiveData
+                                    LiveMatchesListAdapter.ListItemData()
+                            listItemData.mAverageMMR = averageMMR
+                            if (averageMMR < 1)
+                                listItemData.mTitle = "Tournament Match (Match ID " +
```
You can prefix your strings in the xml layout

<textview
src="@string/foobar + etc"
/>

Rhacco replied:
The title is one TextView depending on if it's a tournament or a ranked match. Prefixed strings would not be very convenient in this case

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/viewmodel/MatchesViewModel.kt:
```diff
@@ -6,49 +6,85 @@ import android.arch.lifecycle.MediatorLiveData
+                                listItemData.mTitle = "Tournament Match (Match ID " +
+                                        result.match.matchid + ")"
+                            else
+                                listItemData.mTitle = "Ranked Match (Average MMR " + averageMMR +
```
Same comment as above

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/viewmodel/MatchesViewModel.kt:
```diff
@@ -6,49 +6,85 @@ import android.arch.lifecycle.MediatorLiveData
+                            else
+                                listItemData.mTitle = "Ranked Match (Average MMR " + averageMMR +
+                                        ", Match ID " + result.match.matchid + ")"
+                            if (result.teams.size == 2) {
```
Loop

Rhacco replied:
Fixed

Rhacco committed:
Migrated to RecyclerView

Rhacco committed:
Fixed layouts and namings

Rhacco committed:
Moved data manipulation into the Adapter

Rhacco committed:
Better IDs for players and fixed autoscrolling

Rhacco committed:
Fixed class name

Rhacco committed:
Enhanced comment about how matches are handled

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/sources/remote/Dota2OfficialAPIService.kt:
```diff
@@ -1,18 +1,38 @@
+        if (deviceIsOnline())
+            sDota2OfficialAPIService
+        else
+            null
```
Kotlin, by design, is very averse to null. Rethink this. Avoid null whenever possible and if you do use null, explain why.

Rhacco replied:
Thanks, added a comment

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/viewmodel/MatchesViewModel.kt:
```diff
@@ -0,0 +1,80 @@
+                            val newItemData = LiveMatchesItemData()
+                            newItemData.mAverageMMR = averageMMR
+                            if (averageMMR < 1)
+                                newItemData.mTitle = "Tournament Match (Match ID " +
```
Make Tournament Match and Ranked Match be strings in the strings.xml that take an argument.

Rhacco replied:
Done

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/viewmodel/MatchesViewModel.kt:
```diff
@@ -0,0 +1,80 @@
+                                    result.teams[1].players.size == 5) {
+                                val radiantPlayers = result.teams[0].players
+                                val direPlayers = result.teams[1].players
+                                newItemData.mBlue = radiantPlayers[0].name
```
I am not sure about calling players on radiant and dire by their color names. Color names I think should be considered a visual. Maybe say mRadiantPlayer0, mRadiantPlayer1 and comment the color names in the xml.

I can see why you did it: its within the context of anyone playing dota. But most people who play dota don't remember which color corresponds to raidant or dire and for that reason I don't think it shoudl be used.

Rhacco replied:
I agree, thanks. Done

Rhacco committed:
Improved API service structure and added comment

Rhacco committed:
Moved match strings to resources

Rhacco committed:
Better IDs for player TextViews

Rhacco committed:
Updated screenshot
